draft-ietf-ppm-dap.md

```diff
@@ -1498,30 +1531,57 @@ If the upload request is malformed, the Leader aborts with error
 `unrecognizedTask`.
 
+Otherwise, the Leader responds with a body consisting of an `UploadResponse` and
+the media type `application/dap-upload-resp`. The structure of the response
+is shown below:
+
+~~~ tls-presentation
+struct {
+  ReportID id;
+  ReportError error;
+} ReportUploadStatus;
+
+struct {
+  ReportUploadStatus status[status_count];
+} UploadResponse;
+~~~
+
+The Leader only includes reports that failed processing in the response.
+Reports that are accepted do not have a response.
+
+Reports in the response follow the same order as specified in the request.
+
+For each report that failed to upload, the Leader creates a
+`ReportUploadStatus` struct and includes the `ReportId` from the input and a
+`ReportError` {{basic-definitions}} that describes the failure. The length of
+this sequence is always less than or equal to the length of the upload sequence.
+
 If the Leader does not recognize the `config_id` in the encrypted input share,
-it aborts with an error of type `outdatedConfig`. When the Client receives an
-`outdatedConfig` error, it SHOULD invalidate any cached `HpkeConfigList` and
+it sets the error field to `outdated_config`
+(see {{aggregation-helper-init}}). When the Client receives an
+`outdated_config` error, it SHOULD invalidate any cached `HpkeConfigList` and
 retry with a freshly generated `Report`. If this retried upload does not
 succeed, the Client SHOULD abort and discontinue retrying.
 
 If a report's ID matches that of a previously uploaded report, the Leader MUST
-ignore it. In addition, it MAY alert the Client with error `reportRejected`.
+ignore it. In addition, it MAY alert the Client with setting the
+error field to `report_replayed`.
 
 The Leader MUST ignore any report pertaining to a batch that has already been
-collected (see {{replay-protection}} for details). The Leader MAY also abort
-with error `reportRejected`.
+collected (see {{replay-protection}} for details). The Leader MAY also return
+`report_replayed` error for this report.
 
 The Leader MUST ignore any report whose timestamp is outside of the task's
-`time_interval`. When it does so, it SHOULD also abort with error
-`reportRejected`.
+`time_interval`. When it does so, it SHOULD also set the error field to
+`report_dropped` for the corresponding report in the response.
 
 The Leader may need to buffer reports while waiting to aggregate them (e.g.,
 while waiting for an aggregation parameter from the Collector; see
 {{collect-flow}}). The Leader SHOULD NOT accept reports whose timestamps are too
 far in the future. Implementors MAY provide for some small leeway, usually no
 more than a few minutes, to account for clock skew. If the Leader rejects a
-report for this reason, it SHOULD abort with error `reportTooEarly`. In this
-situation, the Client MAY re-upload the report later on.
+report for this reason, it SHOULD indicate it with `report_too_early` error.
+In this situation, the Client MAY re-upload the report later on.
```

Comment:
Let's use consistent language for error handling to avoid implying that the server should do different things in different cases. "set the error field to `XXX` for the corresponding report in the response" should work